src/sqlfluff/core/dialects/base.py

```diff
@@ -152,7 +152,33 @@ def replace(self, **kwargs: DialectElementType):
         for n in kwargs:
             if n not in self._library:  # pragma: no cover
                 raise ValueError(f"{n!r} is not already registered in {self!r}")
-            self._library[n] = kwargs[n]
+
+            # To replace a segment, the replacement must either be a
+            # subclass of the original, *or* it must have the same
+            # public methods and/or fields as it.
+            cls = kwargs[n]
+            if (
+                isinstance(self._library[n], type)
+                and isinstance(cls, type)
+                and not issubclass(cls, self._library[n])
+            ):
+                if self._library[n].type != cls.type:
+                    raise ValueError(  # pragma: no cover
+                        f"Cannot replace {n!r} because 'type' property does not "
+                        f"match: {cls.type} != {self._library[n].type}"
+                    )
```

Comment:
These lines are actually new. I had wondered if the `type` property was always consistent when replacing a parent segment, and I found several cases where it wasn't. This was presumably a mistake/accident. Several, but not all, of these mismatches were in the Exasol dialect.

Note that when using segment inheritance (now the recommended usual practice when replacing segments), the `type` property can be inherited and thus mismatches avoided.